### PR TITLE
Implementation of node termination handler

### PR DESCRIPTION
**Pull request by vishh**
cc @jiayingz @mindprince 

I have tested this solution manually on PVM and regular GPU clusters. I'm hoping to add GCE specific e2e tests via k8s once this code lands. 
I'd rate this initial commit as alpha quality where it can at-least unblock users. 

Use the [Readme](https://github.com/vishh/k8s-node-termination-handler/blob/86ff045e79b50b62c15c899b94d98a4e3b585d53/README.md) as a guide for reviewing this PR.
It's the [first commit](https://github.com/GoogleCloudPlatform/k8s-node-termination-handler/pull/1/commits/86ff045e79b50b62c15c899b94d98a4e3b585d53) that matters.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# build stage
+FROM golang:latest AS build-env
+RUN go get github.com/golang/dep/cmd/dep
+WORKDIR /go/src/github.com/GoogleCloudPlatform/k8s-node-termination-handler
+RUN git clone https://github.com/GoogleCloudPlatform/k8s-node-termination-handler.git
```
[the remark this answers is not in the document]
Comment (vishh) [reply Dockerfile]: the .build one uses local code. The other one pulls from github to build. The latter is for automating with GCB.

File: _vendor-20180419150601/cloud.google.com/go/.travis.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+sudo: false
```
[the remark this answers is not in the document]
Comment (vishh) [reply _vendor-20180419150601/cloud.google.com/go/.travis.yml]: It is added by dep. Removing from commit history

File: deploy/k8s.yaml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+            - matchExpressions:
+              - key: cloud.google.com/gke-accelerator
+                operator: Exists
+            - matchExpressions:
+              - key: cloud.google.com/gke-preemptible
+                operator: Exists
+      containers:
+      - image: gcr.io/vishnuk-cloud/node-termination-handler@sha256:66f68f10cb552cba66ebadd038c958d275e88b2dba59f242f6ab661fccfec0d7
```
[the remark this answers is not in the document]
Comment (vishh) [reply deploy/k8s.yaml]: Adding a TODO

File: deploy/k8s.yaml [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    type: RollingUpdate
+
+  template:
+    metadata:
+      labels:
+        name: node-termination-handler
+    spec:
+      # Necessary to reboot node
+      hostPID: true
+      # Necessary only when re-using kubelet's Service Accounts is not possible.
```
[the remark this answers is not in the document]
Comment (vishh) [reply deploy/k8s.yaml]: Reusing kubelet's SA provides security by default. Given the recent changes around node authorizer, I suspect we might have to switch the default. Let's leave this issue open for now.

File: termination/types.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	PendingTermination bool
+	// Aboslute time at which the node is expected to be terminated.
+	TerminationTime time.Time
+	// NeedsReboot indicates if a reboot is applicable to handle the pending termination.
+	NeedsReboot bool
+}
+
+// terminationEventSource is an abstract repsentation of objects that tracks impending terminations for a node.
+type NodeTerminationSource interface {
+	// Start launches an internal goroutine that will watch for VM terminations and publish updated to the output channel
```
[the remark this answers is not in the document]
Comment (vishh) [reply termination/types.go]: Updated

File: main.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/GoogleCloudPlatform/k8s-node-termination-handler/termination"
+	"github.com/golang/glog"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
```
[the remark this answers is not in the document]
Comment (vishh) [reply main.go]: Yeah, that's odd. Cleaning it up

File: main.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+const eventSource = "NodeTerminationHandler"
+
+var (
+	inClusterVar        = flag.Bool("in-cluster", false, "Set to false if run outside of a k8s cluster.")
+	regularVMTimeoutVar = flag.Duration("regular-vm-timeout", time.Hour, "Termination timeout for regular VMs. Defaults to an hour which is the timeout duration of GPU VMs.")
+	excludePodsVar      = flag.String("exclude-pods", "", "List of pods to exclude from graceful eviction. Expected format is comma separated 'podName:podNamespace'.")
+	kubeconfig          *string
+	// TODO: Update this to use NoExecute taints once that graduates out of alpha.
+	taintVar                = flag.String("taint", "kubernetes.io/impending-node-termination::NoSchedule", "Taint to place on the node while handling terminations")
```
[the remark this answers is not in the document]
Comment (vishh) [reply main.go]: Good point. Better to use google.com domain for now.

File: main.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/record"
+)
+
+const eventSource = "NodeTerminationHandler"
+
+var (
+	inClusterVar        = flag.Bool("in-cluster", false, "Set to false if run outside of a k8s cluster.")
```
[the remark this answers is not in the document]
Comment (vishh) [reply main.go]: Yeah, I had updated to test locally. Updating the default option

File: Gopkg.lock [hunk 9]
```diff
@@ -0,0 +1,1086 @@
...
+  revision = "75cd24fc2f2c2a2088577d12123ddee5f54e0675"
+
+[[projects]]
+  name = "cloud.google.com/go"
+  packages = ["compute/metadata"]
+  revision = "29f476ffa9c4cd4fd14336b6043090ac1ad76733"
+  version = "v0.21.0"
+
+[[projects]]
+  name = "github.com/Azure/azure-sdk-for-go"
```
[the remark this answers is not in the document]
Comment (vishh) [reply Gopkg.lock]: I have tried vendoring multiple times now. How about we merge this and then you can post a patch to clean it up?

File: main.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+	}
+	excludePods, err := processExcludePods()
+	if err != nil {
+		glog.Fatal(err)
+	}
+	taint, err := processTaint()
+	if err != nil {
+		glog.Fatal(err)
+	}
+	glog.V(10).Infof("Excluding pods %v", excludePods)
```
[the remark this answers is not in the document]
Comment (vishh) [reply main.go]: Good point.

File: termination/taint.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+	nodeTaints := newNode.Spec.Taints
+	if len(nodeTaints) == 0 {
+		return newNode, false, nil
+	}
+
+	if !taintExists(nodeTaints, taint) {
+		return newNode, false, nil
+	}
+
+	newTaints, _ := deleteTaint(nodeTaints, taint)
```
[the remark this answers is not in the document]
Comment (vishh) [reply termination/taint.go]: Smaller methods are easier to read. Also I was trying to replicate existing k8s code which annoyingly isn't public/exported.

File: termination/taint.go [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+	updatedNode, updated, err := removeTaint(node, &n.taint)
+	if err != nil {
+		return err
+	}
+	if updated {
+		if _, err = n.client.CoreV1().Nodes().Update(updatedNode); err != nil {
+			return err
+		}
+		// Log an event that a termination is impending.
+		n.recorder.Eventf(node, v1.EventTypeNormal, untaintReason, "Removing impending termination taint from node %q", n.node)
```
[the remark this answers is not in the document]
Comment (vishh) [reply termination/taint.go]: Good point. We don't need it.

File: termination/taint.go [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+	if err != nil {
+		return err
+	}
+	updatedNode, updated, err := addOrUpdateTaint(node, &n.taint)
+	if err != nil {
+		return err
+	}
+	if updated {
+		if _, err = n.client.CoreV1().Nodes().Update(updatedNode); err != nil {
+			glog.V(10).Infof("Failed to apply taint: %v", err)
```
[the remark this answers is not in the document]
Comment (vishh) [reply termination/taint.go]: updated

File: termination/eviction.go [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+	client               corev1.CoreV1Interface
+	node                 string
+	recorder             record.EventRecorder
+	systemPodGracePeriod time.Duration
+}
+
+// List all pods on the node
+// Evict all pods on the node not in kube-system namespace
+// Return nil on success
+
```
[the remark this answers is not in the document]
Comment (vishh) [reply termination/eviction.go]: typo. removed

File: termination/eviction.go [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+		recorder:             recorder,
+		systemPodGracePeriod: systemPodGracePeriod,
+	}
+}
+
+func (p *podEvictionHandler) EvictPods(excludePods map[string]string, timeout time.Duration) error {
+	options := metav1.ListOptions{FieldSelector: fields.OneTermEqualSelector("spec.nodeName", string(p.node)).String()}
+	pods, err := p.client.Pods(metav1.NamespaceAll).List(options)
+	if err != nil {
+		glog.V(10).Infof("Failed to list pods - %v", err)
```
[the remark this answers is not in the document]
Comment (vishh) [reply termination/eviction.go]: Done. V(2) - error, V(4) - informational

File: main.go [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+	}
+	// creates the clientset
+	return kubernetes.NewForConfig(config)
+}
+
+func homeDir() string {
+	if h := os.Getenv("HOME"); h != "" {
+		return h
+	}
+	return os.Getenv("USERPROFILE") // windows
```
[the remark this answers is not in the document]
Comment (vishh) [reply main.go]: Yeah, removed this method.

**Comment by vishh** [comment]
@mindprince PTAL

File: termination/eviction.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (mindprince) [reply termination/eviction.go]: Should we use `Errorf` for error?

[the remark this answers is not in the document]
Comment (mindprince) [reply termination/eviction.go]: Also, there are still a lot of `V(10)`s left.

```
$ git grep V\(10\) | grep -v vendor
termination/gceTerminationHandler.go:	glog.V(10).Infof("Current states: Regular: %q, PVM: %q", state, pvmState)
termination/handler.go:		glog.V(10).Infof("No pending terminations. Removing taint")
termination/handler.go:	glog.V(10).Infof("Current node state: %v", n.currentNodeState)
termination/handler.go:	glog.V(10).Infof("Applying taint prior to handling termination")
termination/handler.go:	glog.V(10).Infof("Evicting all pods from the node")
termination/handler.go:		glog.V(10).Infof("Rebooting the node")
termination/handler.go:	glog.V(10).Infof("Processing initial node state")
termination/handler.go:		glog.V(10).Infof("Failed to process initial node state - %v", err)
```

File: .travis.yml [hunk 17]
```diff
@@ -0,0 +1,15 @@
+language: go
+
+go:
+  - 1.10
```
Comment (mindprince) [inline .travis.yml]: I think you need to put this in double quotes otherwise travis things this is 1.1

**Comment by vishh** [comment]
@mindprince prow is not enabled for this repo yet. It needs a lgtm via github and a manual merge.

**Pull request merged**